crates/red_knot_python_semantic/src/semantic_index/use_def.rs

```diff
@@ -282,48 +282,61 @@ impl<'db> UseDefMapBuilder<'db> {
         // It's possible we may end up with some old entries in `all_definitions` that nobody is
         // pointing to, but that's OK.
 
-        for (symbol_id, to_merge) in snapshot.definitions_by_symbol.iter_enumerated() {
-            let current = &mut self.definitions_by_symbol[symbol_id];
+        // We never remove symbols from `definitions_by_symbol` (it's an IndexVec, and the symbol
+        // IDs must line up), so the current number of known symbols must always be equal to or
+        // greater than the number of known symbols in a previously-taken snapshot.
+        debug_assert!(self.definitions_by_symbol.len() >= snapshot.definitions_by_symbol.len());
+
+        for (symbol_id, current) in self.definitions_by_symbol.iter_mut_enumerated() {
+            let Some(to_merge) = snapshot.definitions_by_symbol.get(symbol_id) else {
+                // Symbol not present in snapshot, so it's unbound from that path.
+                current.may_be_unbound = true;
+                continue;
+            };
 
             // If the symbol can be unbound in either predecessor, it can be unbound post-merge.
             current.may_be_unbound |= to_merge.may_be_unbound;
 
             // Merge the definition ranges.
-            if current.definitions_range == to_merge.definitions_range {
-                // Ranges already identical, nothing to do!
-            } else if current.definitions_range.end == to_merge.definitions_range.start {
-                // Ranges are adjacent (`current` first), just merge them into one range.
-                current.definitions_range =
-                    (current.definitions_range.start)..(to_merge.definitions_range.end);
-            } else if current.definitions_range.start == to_merge.definitions_range.end {
-                // Ranges are adjacent (`to_merge` first), just merge them into one range.
+            let cur = &mut current.definitions_range;
+            let snap = &to_merge.definitions_range;
+
+            // We never create reversed ranges.
+            debug_assert!(cur.end >= cur.start);
+            debug_assert!(snap.end >= snap.start);
+
+            if cur == snap {
+                // Ranges already identical, nothing to do.
+            } else if snap.is_empty() {
+                // Merging from an empty range; nothing to do.
+            } else if (*cur).is_empty() {
+                // Merging to an empty range; just use the incoming range.
+                current.definitions_range = snap.clone();
```

Comment:
```suggestion
                *cur = snap.clone();
```

The same everywhere else where you assign to the range or don't take a `&mut` for `cur`